### PR TITLE
Create generate free port class to avoid conflicting ports

### LINKED ISSUE #1317: Use random ports in tests
Currently some of our unit tests which need to start some server are using hardcoded ports. This could lead to an issue if the hardcoded port is taken by some other application.

The solution here is to make a small library using `0` port to ask operating system for a random free port. And then use the library everywhere in tests.

### PR DESCRIPTION
**Pull request by PeterChen13579**
Fixes #1317

### CODE REVIEW

**Comment by kuznetsss** [comment]
There are also `tests/unit/web/ServerTests.cpp`, `tests/unit/etl/SubscriptionSourceTests.cpp` and `tests/unit/rpc/ForwardingProxyTests.cpp` where we should use random port.

File: tests/common/util/AssignRandomPort.cpp [hunk 1]
```diff
@@ -0,0 +1,39 @@
+//------------------------------------------------------------------------------
+/*
+    This file is part of clio: https://github.com/XRPLF/clio
+    Copyright (c) 2023, the clio developers.
```
Comment (godexsoft) [inline tests/common/util/AssignRandomPort.cpp]: 2024

File: tests/common/util/AssignRandomPort.hpp [hunk 2]
```diff
@@ -0,0 +1,32 @@
+//------------------------------------------------------------------------------
+/*
+    This file is part of clio: https://github.com/XRPLF/clio
+    Copyright (c) 2023, the clio developers.
```
Comment (godexsoft) [inline tests/common/util/AssignRandomPort.hpp]: also 2024

File: tests/common/util/AssignRandomPort.cpp [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+{
+    boost::asio::io_context io_context;
+    tcp::acceptor acceptor(io_context);
+    tcp::endpoint const endpoint(tcp::v4(), 0);
+
+    acceptor.open(endpoint.protocol());
+    acceptor.set_option(tcp::acceptor::reuse_address(true));
+    acceptor.bind(endpoint);
+
+    tcp::endpoint endPoint = acceptor.local_endpoint();
```
Comment (godexsoft) [inline tests/common/util/AssignRandomPort.cpp]: nit: The names look a bit strange. `endpoint` and then `endPoint`. Consider renaming to something like `localEndpoint` or perhaps we don't even need the temporary there.

File: tests/common/util/AssignRandomPort.hpp [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#pragma once
+
+#include <boost/asio/ts/buffer.hpp>
+#include <boost/asio/ts/internet.hpp>
+
+using tcp = boost::asio::ip::tcp;
```
Comment (godexsoft) [inline tests/common/util/AssignRandomPort.hpp]: This is always a no-no. Never add using in header files, only in cpp files 👍

File: tests/common/util/MockXrpLedgerAPIService.hpp [hunk 5]
```diff
@@ -94,11 +94,17 @@ struct WithMockXrpLedgerAPIService : virtual ::testing::Test {
         serverThread_.join();
     }
 
+    int
+    getXRPLMockPort()
```
Comment (godexsoft) [inline tests/common/util/MockXrpLedgerAPIService.hpp]: This probably should be `const`-qualified.

File: tests/common/util/MockXrpLedgerAPIService.hpp [hunk 6]
```diff
@@ -94,11 +94,17 @@ struct WithMockXrpLedgerAPIService : virtual ::testing::Test {
...
+    {
+        return port_;
+    }
     MockXrpLedgerAPIService mockXrpLedgerAPIService;
 
 private:
     std::unique_ptr<grpc::Server> server_;
     std::thread serverThread_;
+    int port_;
```
Comment (godexsoft) [inline tests/common/util/MockXrpLedgerAPIService.hpp]: Is this the correct type?

**Review by godexsoft — CHANGES_REQUESTED** [review]
Leaving a few comments and suggestions 👍

File: tests/common/util/AssignRandomPort.cpp [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+    boost::asio::io_context io_context;
+    tcp::acceptor acceptor(io_context);
+    tcp::endpoint const endpoint(tcp::v4(), 0);
+
+    acceptor.open(endpoint.protocol());
+    acceptor.set_option(tcp::acceptor::reuse_address(true));
+    acceptor.bind(endpoint);
+
+    tcp::endpoint localEndPoint = acceptor.local_endpoint();
+    return localEndPoint.port();
```
Comment (kuznetsss) [inline tests/common/util/AssignRandomPort.cpp]: Minor thing, you could do
```cpp
return acceptor.local_endpoint().port();
```

File: tests/unit/etl/SubscriptionSourceTests.cpp [hunk 8]
```diff
@@ -46,8 +48,8 @@ struct SubscriptionSourceConnectionTests : public NoLoggerFixture {
     }
 
     boost::asio::io_context ioContext_;
-
-    TestWsServer wsServer_{ioContext_, "0.0.0.0", 11113};
+    uint32_t port = tests::util::generateFreePort();
+    TestWsServer wsServer_{ioContext_, "0.0.0.0", port};
```
Comment (kuznetsss) [inline tests/unit/etl/SubscriptionSourceTests.cpp]: I think instead it is better to remove port parameter from `TestWsServer`'s constructor and use always `0`.  Then add `port()` method for `TestWsServer` and use it below.

File: tests/unit/util/requests/RequestBuilderTests.cpp [hunk 9]
```diff
@@ -50,8 +52,9 @@ struct RequestBuilderTestBundle {
 };
 
 struct RequestBuilderTestBase : SyncAsioContextTest {
-    TestHttpServer server{ctx, "0.0.0.0", 11111};
-    RequestBuilder builder{"localhost", "11111"};
+    uint32_t port{tests::util::generateFreePort()};
+    TestHttpServer server{ctx, "0.0.0.0", port};
+    RequestBuilder builder{"localhost", std::to_string(port)};
```
Comment (kuznetsss) [inline tests/unit/util/requests/RequestBuilderTests.cpp]: And the same as for `TestWsServer` here.

File: tests/unit/web/ServerTests.cpp [hunk 10]
```diff
@@ -166,14 +182,21 @@ class WebServerTest : public NoLoggerFixture {
...
+    std::string
+    portToString() const
+    {
+        return std::to_string(port);
+    }
+
     // this ctx is for dos timer
     boost::asio::io_context ctxSync;
-    Config cfg{boost::json::parse(JSONData)};
+    uint32_t port = tests::util::generateFreePort();
```
Comment (kuznetsss) [inline tests/unit/web/ServerTests.cpp]: Probably better to save it as `std::string`

File: tests/common/util/TestHttpServer.cpp [hunk 11]
```diff
@@ -134,3 +134,9 @@ TestHttpServer::handleRequest(TestHttpServer::RequestHandler handler, bool const
         boost::asio::detached
     );
 }
+
+int
+TestHttpServer::port() const
```
Comment (kuznetsss) [inline tests/common/util/TestHttpServer.cpp:139]: You also should remove `port` parameter form constructor for `TestHttpServer` and `TestWsServer` and use `0` instead.

**Review by kuznetsss — DISMISSED** [review]
👍

File: tests/common/util/TestHttpServer.hpp [hunk 12]
```diff
@@ -43,7 +43,7 @@ class TestHttpServer {
      * @param host host to bind to
      * @param port port to bind to
```
Comment (godexsoft) [inline tests/common/util/TestHttpServer.hpp]: This will later trigger doxygen to error out.. best removed now since the param is already gone 👍

**Review by godexsoft — CHANGES_REQUESTED** [review]
Looks good, just fix the doxy comment and let's ship it 🚀

**Review by godexsoft — DISMISSED** [review]

**Review by kuznetsss — APPROVED** [review]

**Pull request merged**